Add onreply_route async capabilities

vladpaiu opened the pull request:
**Summary**
Add onreply_route async capabilities

**Solution**
Clone reply before running async functions & free afterwards

**Compatibility**
Backwards incompatible, previous async in onreply_route was run in sync mode.

bogdan-iancu commented:
@vladpaiu , this sounds awesome !!! Let me check it !

bogdan-iancu commented on modules/tm/t_reply.c:
```diff
@@ -1671,6 +1674,19 @@ int reply_received( struct sip_msg  *p_msg )
+		if (async_status > 0) {
+			/* async was started in the onreply route, no need to do anything more here */
+			/* mark that the UAC received replies */
+			uac->flags |= T_UAC_HAS_RECV_REPLY;
```
does it make any difference if we move the setting of the flag before running the onreply route, so we do not have to duplicate ?

vladpaiu replied:
4 hours later it struck me that you were talking about T_UAC_HAS_RECV_REPLY :))

to be honest, I'm not sure if it matters or not, but to keep things consistent as they were before, I'm setting it after the reply route has been run ( with async or without async )

bogdan-iancu replied:
it makes sense, let's keep it like this for now

bogdan-iancu commented on modules/tm/async.c:
```diff
@@ -319,8 +553,18 @@ int t_handle_async(struct sip_msg *msg, struct action* a,
 	}
 
 	if (route_type!=REQUEST_ROUTE) {
```
Some
```
if ()
else if ()
else
```
may look better here :)

vladpaiu replied:
fixed in b2811a5f4c8403695acd7bdb92a7fe1f3df87881

resolving comment

bogdan-iancu commented on modules/tm/async.c:
```diff
@@ -219,6 +217,242 @@ int t_resume_async(int fd, void *param, int was_timeout)
+	/* DO TM suspended actions to decide if we need to relay */
+	LOCK_REPLIES( t );
+
+	/* we fire a cancel on spot if (a) branch is marked "to be canceled" or (b)
```
this chunk below is duplicated from the `reply_received()` (all the way to `restore_exit_reply` label). I haven;t looked into details, but maybe we should avoid code duplicate, especially in such a delicate section, on reply handing. Maybe some refactoring into some new inline function ?

vladpaiu replied:
address in 7477f3f42b669f6c8dcfb0f0c1af73d1e84706ab , resolving comment

bogdan-iancu commented on modules/tm/async.c:
```diff
@@ -219,6 +217,242 @@ int t_resume_async(int fd, void *param, int was_timeout)
+	/* run the resume_route (some type as the original one) */
+	swap_route_type(route, ctx->route_type);
+	/* do not run any post script callback, we are a reply */
+	run_resume_route( ctx->resume_route, ctx->reply, 0);
```
The initial onreply route allows you to run with the lock (see the `onreply_avp_mode` setting) or without lock. The resume route is not impacted (correct me if I'm wrong) by this setting, leading to concurrent access to the transactional list of AVPS.

vladpaiu replied:
good catch ! fixed under dd79818af2ee0c10d1b37c0f1a648888cb54e302, resolving comment

bogdan-iancu commented on modules/tm/async.c:
```diff
@@ -219,6 +217,242 @@ int t_resume_async(int fd, void *param, int was_timeout)
+	set_cancelled_t(ctx->cancelled_t);
+	set_e2eack_t(ctx->e2eack_t);
+	reset_kr();
+	set_kr(ctx->kr);
```
I do not see this `ctx->kr` set anywhere (in the new code). Do we need this? KR is typical for request processing, not for replies.
Similar the e2eack and cancel stuff.

vladpaiu replied:
addressed in 16ef67d27e1fabec487d65b41f1b502247f3093f , resolving comment

vladpaiu commented:
@bogdan-iancu addressed some of the comments, also fixed a mem leak under https://github.com/OpenSIPS/opensips/pull/3550/commits/6811373f7791e99749d7fb1cced36c75a6604e7f that I saw today while doing a bit of load testing to make sure all's good after recent changes.

vladpaiu commented:
@bogdan-iancu  any other things to address before we can merge this ?

bogdan-iancu commented:
I'm on it :)
